lib/graphql/static_validation/rules/fields_will_merge.rb

```diff
@@ -239,13 +232,50 @@ def find_conflict(response_key, field1, field2, mutually_exclusive: false)
           end
         end
 
+        if !conflicts[:field].key?(response_key) &&
+            (t1 = field1.definition&.type) &&
+            (t2 = field2.definition&.type) &&
+            return_types_conflict?(t1, t2)
+
+          conflict = conflicts[:return_type][response_key]
+          conflict.add_conflict(node1, "`#{t1.to_type_signature}`")
+          conflict.add_conflict(node2, "`#{t2.to_type_signature}`")
+
+          @conflict_count += 1
+        end
+
         find_conflicts_between_sub_selection_sets(
           field1,
           field2,
           mutually_exclusive: are_mutually_exclusive,
         )
       end
 
+      def return_types_conflict?(type1, type2)
+        if type1.list?
+          if type2.list?
+            return_types_conflict?(type1.of_type, type2.of_type)
+          else
+            true
+          end
+        elsif type2.list?
```

Comment:
Comparing this with NodeJS... https://github.com/graphql/graphql-js/blob/main/src/validation/rules/OverlappingFieldsCanBeMergedRule.ts#L694-L718. Shouldn't this return `true`?